### PR TITLE
[Feature] Support BigGAN and BigGAN-deep

### PR DESCRIPTION
**Pull request by plyfager**
Add generator's architecture of BigGAN.

### CODE REVIEW

File: mmgen/models/architectures/biggan/modules.py [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+                 shortcut_cfg=dict(type='Conv2d'),
+                 act_cfg=dict(type='ReLU'),
+                 upsample_cfg=dict(type='nearest', scale_factor=2),
+                 with_spectral_norm=True,
+                 shared_embedding=True,
+                 auto_sync_bn=True):
+
+        super().__init__()
+        self.activation = build_activation_layer(act_cfg)
+        self.upsample_cfg = upsample_cfg
```
Comment (nbei) [inline mmgen/models/architectures/biggan/modules.py:31]: `self.upsample_cfg = deepcopy(upsample_cfg)`

File: mmgen/models/architectures/biggan/modules.py [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+    def __init__(self,
+                 num_features,
+                 input_dim,
+                 bn_eps=1e-5,
+                 sn_eps=1e-6,
+                 momentum=0.1,
+                 shared_embedding=True,
+                 with_spectral_norm=True,
+                 auto_sync_bn=True):
+        super(BigGANConditionBatchNorm, self).__init__()
```
Comment (nbei) [inline mmgen/models/architectures/biggan/modules.py:107]: `super().__init__()`

File: mmgen/models/architectures/biggan/modules.py [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+
+        self.bn = nn.BatchNorm2d(
+            num_features, eps=bn_eps, momentum=momentum, affine=False)
+
+        if auto_sync_bn and dist.is_initialized():
+            self.bn = SyncBatchNorm.convert_sync_batchnorm(self.bn)
+
+    def forward(self, x, y):
+        # Calculate class-conditional gains and biases
+        gain = (1 + self.gain(y)).view(y.size(0), -1, 1, 1)
```
Comment (nbei) [inline mmgen/models/architectures/biggan/modules.py:130]: `(1. + self.gain(y))`

File: mmgen/models/architectures/biggan/modules.py [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+        out = self.bn(x)
+        return out * gain + bias
+
+
+@MODULES.register_module()
+class SelfAttentionBlock(nn.Module):
+
+    def __init__(self,
+                 in_channels,
+                 conv_cfg=dict(type='Conv2d'),
```
Comment (nbei) [inline mmgen/models/architectures/biggan/modules.py:141]: This `conv_cfg` is only for the `conv_cfg` in `ConvModule`. However, we hardly change this in this `ConvModule`. Because we only adopt the standard convolution here. Thus, I may recommend removing this config. Otherwise, you should use conv_cfg like:

```python
self.theta = ConvModule(
    self.in_channels,
    self.in_channels // 8,
    kernel_size=1,
    bias=False,
    act_cfg=None,
    padding=0,
    **conv_cfg)
```

File: mmgen/models/architectures/biggan/generator_discriminator.py [hunk 5]
```diff
@@ -0,0 +1,532 @@
...
+                    head_block=(index == 0)))
+            self.conv_blocks.append(build_module(self.blocks_cfg))
+            if self.arch['attention'][index]:
+                self.conv_blocks.append(
+                    SelfAttentionBlock(
+                        self.arch['out_channels'][index],
+                        with_spectral_norm=with_spectral_norm,
+                        sn_eps=sn_eps))
+
+        self.activation = build_activation_layer(act_cfg)
```
Comment (nbei) [inline mmgen/models/architectures/biggan/generator_discriminator.py:416]: `self.activate` is better

File: mmgen/models/architectures/biggan/generator_discriminator.py [hunk 6]
```diff
@@ -0,0 +1,532 @@
...
+        self.blocks_cfg = deepcopy(blocks_cfg)
+        self.blocks_cfg.update(
+            dict(
+                conv_cfg=conv_cfg,
+                act_cfg=act_cfg,
+                sn_eps=sn_eps,
+                with_spectral_norm=with_spectral_norm))
+
+        self.conv_blocks = nn.ModuleList()
+        for index in range(len(self.arch['out_channels'])):
```
Comment (nbei) [inline mmgen/models/architectures/biggan/generator_discriminator.py:400]: Similar to the comment in generator.

File: mmgen/models/architectures/biggan/generator_discriminator.py [hunk 7]
```diff
@@ -0,0 +1,532 @@
...
+            if self.arch['attention'][index]:
+                self.conv_blocks.append(
+                    SelfAttentionBlock(
+                        self.arch['out_channels'][index],
+                        with_spectral_norm=with_spectral_norm,
+                        sn_eps=sn_eps))
+
+        self.activation = build_activation_layer(act_cfg)
+
+        self.linear = nn.Linear(self.arch['out_channels'][-1], out_channels)
```
Comment (nbei) [inline mmgen/models/architectures/biggan/generator_discriminator.py:418]: `self.decision_layer` may be better.

File: mmgen/models/architectures/biggan/modules.py [hunk 8]
```diff
@@ -0,0 +1,416 @@
...
+                 bn_eps=1e-5,
+                 sn_eps=1e-6,
+                 momentum=0.1,
+                 label_input=False,
+                 with_spectral_norm=True,
+                 auto_sync_bn=True):
+        super().__init__()
+        assert num_features > 0 and input_dim > 0
+        # Prepare gain and bias layers
+        if not label_input:
```
Comment (nbei) [inline mmgen/models/architectures/biggan/modules.py:186]: Please specify the design of the label in this part.  What's the relation between `input_dim` and `num_features`. You should also consider changing the variable name in this module.

File: mmgen/models/architectures/biggan/modules.py [hunk 9]
```diff
@@ -0,0 +1,416 @@
...
+            self.gain = nn.Linear(input_dim, num_features, bias=False)
```
Comment (nbei) [inline mmgen/models/architectures/biggan/modules.py:187]: What's the meaning of `input_dim`.

File: mmgen/models/architectures/biggan/modules.py [hunk 10]
```diff
@@ -0,0 +1,416 @@
...
+        self,
+        in_channels,
+        out_channels,
+        conv_cfg=dict(type='Conv2d'),
+        shortcut_cfg=dict(type='Conv2d'),
+        act_cfg=dict(type='ReLU', inplace=False),
+        sn_eps=1e-6,
+        with_downsample=True,
+        with_spectral_norm=True,
+        head_block=False,
```
Comment (nbei) [inline mmgen/models/architectures/biggan/modules.py:333]: remove the comma here.

File: mmgen/models/architectures/biggan/modules.py [hunk 11]
```diff
@@ -0,0 +1,416 @@
...
+    ):
+        super().__init__()
+        self.activation = build_activation_layer(act_cfg)
+        self.with_downsample = with_downsample
+        self.head_block = head_block
```
Comment (nbei) [inline mmgen/models/architectures/biggan/modules.py:338]: `is_head_block`? What's the meaning of `head _block`. In the forward function, I guess that you use this variable to control the order of modules?

**Comment by nbei** [comment]
The unit test can be further added for more coverage.

File: mmgen/models/architectures/biggan/generator_discriminator.py [hunk 12]
```diff
@@ -0,0 +1,532 @@
...
+    def init_weights(self, pretrained=None, init_type='ortho'):
+        """Init weights for models.
+
+        We just use the initialization method proposed in the original paper.
+
+        Args:
+            pretrained (str, optional): Path for pretrained weights. If given
+                None, pretrained weights will not be loaded. Defaults to None.
+        """
+        if isinstance(pretrained, str):
```
Comment (nbei) [inline mmgen/models/architectures/biggan/generator_discriminator.py:322]: We may support load pre-trained model like this:
https://github.com/open-mmlab/mmgeneration/blob/master/mmgen/models/architectures/stylegan/generator_discriminator_v2.py#L211

Because there are many works using pretrained BigGAN generator and BigGAN discriminator.

File: mmgen/models/architectures/biggan/generator_discriminator.py [hunk 13]
```diff
@@ -0,0 +1,532 @@
...
+        """Init weights for models.
+
+        We just use the initialization method proposed in the original paper.
+
+        Args:
+            pretrained (str, optional): Path for pretrained weights. If given
+                None, pretrained weights will not be loaded. Defaults to None.
+        """
+
+        if isinstance(pretrained, str):
```
Comment (nbei) [inline mmgen/models/architectures/biggan/generator_discriminator.py:514]: We may support load pre-trained model like this:
https://github.com/open-mmlab/mmgeneration/blob/master/mmgen/models/architectures/stylegan/generator_discriminator_v2.py#L211

Because there are many works using pretrained BigGAN generator and BigGAN discriminator.

**Review by nbei — COMMENTED** [review]
Pre-trained BigGAN models.

File: mmgen/models/architectures/biggan/generator_discriminator.py [hunk 14]
```diff
@@ -0,0 +1,532 @@
...
+            self.noise_size // self.num_slots,
+            self.arch['in_channels'][0] * (self.input_scale**2))
+        if with_spectral_norm:
+            self.noise2feat = spectral_norm(self.noise2feat, eps=sn_eps)
+
+        # If using 'shared_embedding', we will get an unified embedding of
+        # label for all blocks. If not, we just pass on the label to each
+        # block.
+        if with_shared_embedding:
+            self.shared_embedding = nn.Embedding(num_classes, shared_dim)
```
[the remark this answers is not in the document]
Comment (plyfager) [reply mmgen/models/architectures/biggan/generator_discriminator.py:122]: In `Pytorch-StudioGAN` and `BigGAN-PyTorch`, this shared embedding does not need spectral norm.

File: mmgen/models/architectures/biggan/generator_discriminator.py [hunk 15]
```diff
@@ -0,0 +1,532 @@
...
+                [base_channels * item for item in [16, 8, 8, 4, 2, 1, 1]],
+                'upsample': [True] * 7,
+                'resolution': [8, 16, 32, 64, 128, 256, 512],
+                'attention': [False, False, False, True, False, False, False]
+            }
+        }
+
+        return _default_arch_cfgs[str(output_scale)]
+
+    def forward(self, noise, label, num_batches=0, return_noise=False):
```
[the remark this answers is not in the document]
Comment (plyfager) [reply mmgen/models/architectures/biggan/generator_discriminator.py:225]: Alright, I'll fix this.

File: mmgen/models/architectures/biggan/generator_discriminator.py [hunk 16]
```diff
@@ -0,0 +1,532 @@
...
+            label_batch = None
+
+        elif isinstance(label, torch.Tensor):
+            assert label.ndim == 1, ('The label shoube be in shape of (n, )'
+                                     f'but got {label.shape}.')
+            label_batch = label
+        elif callable(label):
+            label_generator = label
+            assert num_batches > 0
+            label_batch = label_generator((num_batches))
```
[the remark this answers is not in the document]
Comment (plyfager) [reply mmgen/models/architectures/biggan/generator_discriminator.py:264]: Alright, I'll fix this.

File: mmgen/models/architectures/biggan/generator_discriminator.py [hunk 17]
```diff
@@ -0,0 +1,532 @@
...
+        self.arch = arch_cfg if arch_cfg else self._get_default_arch_cfg(
+            self.output_scale, base_channels)
+        self.input_scale = input_scale
+        self.split_noise = split_noise
+        self.blocks_cfg = deepcopy(blocks_cfg)
+        self.upsample_cfg = deepcopy(upsample_cfg)
+
+        # Validity Check
+        # If 'num_classes' equals to zero, we shall set 'with_shared_embedding'
+        # to False and
```
[the remark this answers is not in the document]
Comment (plyfager) [reply mmgen/models/architectures/biggan/generator_discriminator.py:90]: Well, in default settings, if users set `num_classes` to zero, they should set `split_noise` to True since we use `cbn`. But if users do not use `cbn`,  `split_noise` may be False. I am wondering how to write the comments.

File: mmgen/models/architectures/biggan/generator_discriminator.py [hunk 18]
```diff
@@ -0,0 +1,532 @@
...
+
+        # First linear layer
+        self.noise2feat = nn.Linear(
+            self.noise_size // self.num_slots,
+            self.arch['in_channels'][0] * (self.input_scale**2))
+        if with_spectral_norm:
+            self.noise2feat = spectral_norm(self.noise2feat, eps=sn_eps)
+
+        # If using 'shared_embedding', we will get an unified embedding of
+        # label for all blocks. If not, we just pass on the label to each
```
[the remark this answers is not in the document]
Comment (plyfager) [reply mmgen/models/architectures/biggan/generator_discriminator.py:119]: I will just use `pass`.

File: mmgen/models/architectures/biggan/generator_discriminator.py [hunk 19]
```diff
@@ -0,0 +1,532 @@
...
+                dim_after_concat=self.dim_after_concat,
+                act_cfg=act_cfg,
+                conv_cfg=conv_cfg,
+                sn_eps=sn_eps,
+                label_input=(num_classes > 0) and (not with_shared_embedding),
+                with_spectral_norm=with_spectral_norm,
+                auto_sync_bn=auto_sync_bn))
+
+        self.conv_blocks = nn.ModuleList()
+        for index in range(len(self.arch['out_channels'])):
```
[the remark this answers is not in the document]
Comment (plyfager) [reply mmgen/models/architectures/biggan/generator_discriminator.py:144]: Alright, I'll take this advice.

File: mmgen/models/architectures/biggan/generator_discriminator.py (hunk 7 above)
Comment (plyfager) [reply mmgen/models/architectures/biggan/generator_discriminator.py:418]: I will just use `self.decision`.

File: mmgen/models/architectures/biggan/modules.py (hunk 9 above)
Comment (plyfager) [reply mmgen/models/architectures/biggan/modules.py:187]: The dim of linear layer's input in `BigGANConditionBN`. If with shared embedding, the input is z concatenated with class vector. If not, the input is the label.
